### PR TITLE
[app_dart] Migrate to null safety

### PR DESCRIPTION
**Pull request by CaseyHillers**
Fixes https://github.com/flutter/flutter/issues/82705
Fixes https://github.com/flutter/flutter/issues/88185 - ci is no longer using `flutter upgrade`
Fixes https://github.com/flutter/flutter/issues/89709

### Cocoon

* Remove `CirrusGraphQLClient`
  * The same as `GraphQLClient`
* Migrate BuildBucket id from int to string
  * This is because json_serializable now exports it as a JSON number which is not safe + there's no use of it as an int
* Created test entity_generators.dart with common utility functions

#### Fakes
- FakeRequestHandler
- FakeBigqueryService (rename of BigqueryServiceMock)

#### Mocks

* Null safety requires generated the mock model code (this is under test/src/utilities/mocks.dart)
* Any tests which were able to assume a null value returned were updated to include a stub

### google_apis
* Api dropped from various classes

### GraphQL

* Add lower level library to enable access to moved datastructures
* Update `QueryResult` errors to the renamed exceptions
  * Add `source`in tests
* Parse GraphQL strings to `DocumentNode` 

### http
* Move string urls to `Uri`
* Encourage use of `package:http`

### metrics_center
* Add commit timestamp

## Test Plan
- [x] `dart test` passes
- [x] Local server passes all usable functionality (serves content, views dashboard, retries tasks)
- [x] 1% prod traffic test
- [x] 10% prod traffic test
- [x] 25% prod traffic test
- [x] 50% prod traffic test
- [ ] Full roll out

## Preview

https://testchillers-dot-flutter-dashboard.appspot.com

### CODE REVIEW

File: app_dart/lib/src/request_handlers/luci_status.dart
Comment (CaseyHillers) [inline app_dart/lib/src/request_handlers/luci_status.dart:76]: I'm not sure why this is now surfacing as an issue in the new Dart version.

`envelope.message.data` has always been a base64 encoded string of the json, however it seems json decode removed the base64 decode aspect

**Comment by CaseyHillers** [comment]
There's 21 tests failures remaining that require non-trivial fixes. 

- 6 have to do with the flaky issues using a generic function on a mock, which isn't supported (it was able to rely on a null value)
- 9 have to do with some weird mockito setup and not using a provided stub (github webhook)
- 6 have to do with previously null bigquery values in refresh chromebot status

**Comment by godofredoc** [comment]
> There's 21 tests failures remaining that require non-trivial fixes.
> 
> * 6 have to do with the flaky issues using a generic function on a mock, which isn't supported (it was able to rely on a null value)
> * 9 have to do with some weird mockito setup and not using a provided stub (github webhook)
> * 6 have to do with previously null bigquery values in refresh chromebot status

Is this something that we can crowdsource? it is blocking already several changes to the backend.

**Comment by CaseyHillers** [comment]
That would be great as I need to head out for an appointment. Feel free to add commits to this PR for the fixes.

The remaining handlers can be crowdsourced:

1. get-status
2. refresh-chromebot-status
3. github-webhook
4. file-flaky-issue
5. Add mocks.mocks.dart to the license check

Once these are done, we'll need to squash the PR into one commit, and rebase with tip of tree, fixing the issues there.

Then we can format and do a prod test.

**Comment by CaseyHillers** [comment]
There's one other issue, and I'm trying to find a solution:

```dart
when(mock.methodThatReturnsStream).thenAnswer((_) => Stream.empty());
```

Mockito will not return `Stream.emtpy()` in the tests, or store it as a stub. I asked in #hackers-test to get ideas, but from the documentation it sounds like this should work.

**Comment by CaseyHillers** [comment]
I'm working on the last issue regarding the mock generics. I was able to find https://github.com/dart-lang/mockito/blob/master/NULL_SAFETY_README.md#fallback-generators and i'm working to implement it now.

**Comment by CaseyHillers** [comment]
I'm disabling the flaky issue tests as they're non-critical and blocking this PR. https://github.com/flutter/flutter/issues/90139

**Comment by CaseyHillers** [comment]
I'm going to squash these commits to rebase on top of master (\cc @keyonghan @godofredoc feel free to review now)

**Review by keyonghan — COMMENTED** [review]
I wouldn't expect this to break main features, but can we validate the PR in a test version before merging?

**Comment by CaseyHillers** [comment]
> I wouldn't expect this to break main features, but can we validate the PR in a test version before merging?

Yes. My expectation is the review will take longer than the prod traffic test, so I wanted to get early eyes on it. I'm also debugging issues through the local dart server

**Comment by CaseyHillers** [comment]
Another AI is I need to revisit the model classes and mark everything as nullable, otherwise dart:mirrors doesn't work properly

**Comment by CaseyHillers** [comment]
The only issue i'm noticing is `BuildBucketClient._postRequest` is always failing with `BuildBucketException: [400]: could not decode body: illegal base64 data at input byte 0`.

I'm adding some logs to see what the message looks like, and see if there are any missing headers.

**Comment by CaseyHillers** [comment]
> The only issue i'm noticing is BuildBucketClient._postRequest is always failing with BuildBucketException: [400]: could not decode body: illegal base64 data at input byte 0.

After some more debugging, it's not always failing. For some reason it can pass on the exact same call (like refresh-chromebot-status querying all the same builds, but then fail on push-build-status-to-github)

File: app_dart/lib/src/service/scheduler.dart
Comment (CaseyHillers) [inline app_dart/lib/src/service/scheduler.dart:384]: I increased ci.yaml check failures to warnings to surface issues in the logs

**Comment by CaseyHillers** [comment]
I did a prod traffic test with service/buildbucket.dart using the original dart:io calls, and that is still returning a malformed response. I suspect there is an issue in the model/buildbucket.dart migration

File: app_dart/lib/src/service/luci_build_service.dart
Comment (godofredoc) [inline app_dart/lib/src/service/luci_build_service.dart:345]: You'll need to transform the encoded json string to base64.

Comment (CaseyHillers) [reply app_dart/lib/src/service/luci_build_service.dart:345]: This was it, thank you! I'm not sure why this regressed with this PR

**Review by godofredoc — APPROVED** [review]

**Comment by CaseyHillers** [comment]
I'm submitting this, but there's a known issue where the waiting for tree to green graphql queries can timeout ~20% of the time.  I filed to https://github.com/flutter/flutter/issues/90640 to track

**Pull request merged**
